### PR TITLE
pkg/metrics/bpf: new bpf_maps & bpf_progs metrics

### PR DESCRIPTION
**Pull request by mvisonneau**
This change introduces two new metrics to monitor the amount of bpf maps and programs currently defined on the system.

We recently discovered that abnormally large amounts of maps could be found on some of our nodes.

Additionally, this change introduces request concurrency limit as well as dynamic definition of the metrics list throughout the Describe() function.

```release-note
pkg/metrics/bpf: new bpf_maps & bpf_progs metrics
```

### CODE REVIEW

File: pkg/metrics/bpf.go
Comment (derailed) [inline pkg/metrics/bpf.go]: Not sure if this is necessary? Should this be called only per scrape interval?

**Review by derailed — COMMENTED** [review]
@mvisonneau LGTM Thank you for this update!

File: pkg/metrics/bpf.go
Comment (mvisonneau) [reply pkg/metrics/bpf.go]: Yeah, I've added it to prevent a potential DoS attack over the metrics endpoint. IIUC this is being called on every `GET /metrics` requests, isn't it? 🤔 

Happy to remove it if you feel this is overkill!

Comment (ti-mo) [reply pkg/metrics/bpf.go]: Is Collect() invoked on each hit to /metrics? Or does the agent cache these metrics somewhere and periodically refresh them?

**Comment by mvisonneau** [comment]
IIUC, it is indeed invoked on each /metrics query

**Comment by mvisonneau** [comment]
👋 hey @ti-mo, indeed my bad 🙈 , I've rebased and hopefully have also sorted all failing tests!

**Comment by ldelossa** [comment]
@mvisonneau 

There seems to be two consistently failing tests here: 
https://github.com/cilium/cilium/actions/runs/8003368791?pr=29984
https://github.com/cilium/cilium/actions/runs/8114768581
https://github.com/cilium/cilium/actions/runs/8114769384

If you feel your PR is not the cause of these please attempt a rebase and we will run tests again.

**Comment by joestringer** [comment]
The smoke test failures report an issue with the new metrics in the PR:

```
cilium_bpf_maps_count non-histogram and non-summary metrics should not have "_count" suffix
cilium_bpf_progs_count non-histogram and non-summary metrics should not have "_count" suffix
Error: Process completed with exit code 3.
```

File: pkg/metrics/bpf.go
Comment (chancez) [inline pkg/metrics/bpf.go]: Might be a good idea to return a struct instead of the 2 `uint64`, to make it more obvious what each value corresponds to.

File: pkg/metrics/bpf.go
Comment (chancez) [reply pkg/metrics/bpf.go]: @ti-mo `Collect` is called on every scrape. For "normal" metrics, that's just reading a bunch of atomics, so it's not a big deal, for this case, I'll defer to others if this is needed. Another alternative to a simple mutex is to leverage the https://pkg.go.dev/golang.org/x/sync/singleflight package.

Comment (ti-mo) [reply pkg/metrics/bpf.go]: Yep, singleflight is neat for this use case!

File: pkg/metrics/bpf.go
Comment (mvisonneau) [reply pkg/metrics/bpf.go]: sounds good indeed, I've added a new `bpfUsage` struct to support those variables.

File: pkg/metrics/bpf.go
Comment (mvisonneau) [reply pkg/metrics/bpf.go]: Thanks for making me learn about this nice feat. I've replaced my former mutex implementation with it but not sure if it makes it easier in the end 🤷 let me know if there are better ways to integrate it!

**Comment by chancez** [comment]
> The smoke test failures report an issue with the new metrics in the PR:
> 
> ```
> cilium_bpf_maps_count non-histogram and non-summary metrics should not have "_count" suffix
> cilium_bpf_progs_count non-histogram and non-summary metrics should not have "_count" suffix
> Error: Process completed with exit code 3.
> ```
> 
> (Maybe `_total` is a better suffix? @cilium/metrics folks could provide better guidance, but this is using standard prometheus tooling to validate the naming)

https://prometheus.io/docs/practices/naming/ and https://www.robustperception.io/on-the-naming-of-things/ are a good reference. Generally `_total` is for accumulating counts, so I don't know that `_total` is appropriate here. I think you would probably just leave it as `cilium_bpf_maps` and `cilium_bpf_progs` (the unit is maps and programs).

**Comment by mvisonneau** [comment]
Thanks for your feedback @chancez, I have looked into addressing your suggestions!

File: pkg/metrics/bpf.go
Comment (chancez) [inline pkg/metrics/bpf.go]: I know I suggested single flight, but I think this is an incorrect usage. You need to use the return values, otherwise only the first caller of `Collect` is going to get the values. Single flight basically acts as a concurrent memoizer.

**Review by chancez — CHANGES_REQUESTED** [review]
I believe the usage of single flight is incorrect and needs fixing, otherwise when there are multiple concurrent calls to collect, only 1 will actually receive the metrics.

File: pkg/metrics/bpf.go
Comment (mvisonneau) [reply pkg/metrics/bpf.go]: Isn't it supposed to be blocking? 🤔 https://pkg.go.dev/golang.org/x/sync/singleflight#Group.Do
What should we be doing with the returned values?

Comment (mvisonneau) [reply pkg/metrics/bpf.go]: ohhh I see 🙈

Comment (mvisonneau) [reply pkg/metrics/bpf.go]: alright, I've made some updates which should give more sense to the implementation of the singleflight lib!

File: pkg/metrics/bpf.go
Comment (chancez) [inline pkg/metrics/bpf.go:145]: This should probably return if there's an error. Previously if `getMemoryUsage` failed, it didn't send the value on `ch`.

**Review by chancez — CHANGES_REQUESTED** [review]
Based on the code previously, the metrics weren't updated if there was an error, but this logs the error and updates the metrics anyways, which means they'll probably get set to 0, which doesn't feel right to me.

**Review by tklauser — APPROVED** [review]
LGTM for @cilium/vendor

**Comment by tklauser** [comment]
@mvisonneau there still seem to be some unaddressed review comments. Are you still working on this PR?

File: pkg/metrics/bpf.go
Comment (mvisonneau) [reply pkg/metrics/bpf.go:145]: absolutely, I've reverted the change 👍

**Comment by mvisonneau** [comment]
Apologies, I missed out on those last comments! They should be addressed now and I've also rebased the branch 👍

**Review by chancez — APPROVED** [review]

File: pkg/metrics/bpf.go
Comment (tommyp1ckles) [inline pkg/metrics/bpf.go]: Your changes are consistent with how things where done before, but I find it a bit unexpected that we don't only count maps prefixed with `cilium_`

File: pkg/metrics/bpf.go
Comment (tommyp1ckles) [inline pkg/metrics/bpf.go]: nit: we could do this type assertion just once to make thing less verbose.

**Review by tommyp1ckles — APPROVED** [review]
Thanks @mvisonneau , I think this is a worthwhile metric to add.  Changes look good to me.

Looks like there's some conflicts in the vendor file that need to be resolved.

**Comment by ti-mo** [comment]
@mvisonneau Ping, this needs a rebase

File: pkg/metrics/bpf.go
Comment (mvisonneau) [reply pkg/metrics/bpf.go]: do you feel we should we change it as part of this PR or through a separate one though?

File: pkg/metrics/bpf.go
Comment (mvisonneau) [reply pkg/metrics/bpf.go]: definitely, I've updated this implementation 👍

**Comment by mvisonneau** [comment]
my bad, I once again missed the notifications. I've rebased and looked into one of @tommyp1ckles's suggestion. Let me know if you want us to scope down to maps prefixed with `cilium_` as part of this PR.

**Comment by ti-mo** [comment]
> Let me know if you want us to scope down to maps prefixed with `cilium_` as part of this PR.

I think that would be wise. There's more and more bpf tooling out there nowadays, so we'd report on the whole bpf subsystem's memory consumption, which would be quite surprising for a user that expects this number to be accurate. bpftool itself creates maps during its runtime, also things like systemd, etc.

**Comment by mvisonneau** [comment]
sounds good, I've added some logic to filter on **maps** prefixed with `cilium_` and **programs** related to cilium maps 👍

**Comment by aanm** [comment]
@mvisonneau it seems that the CI is failing:
```
  Running [/home/runner/golangci-lint-1.57.2-linux-amd64/golangci-lint run --out-format=github-actions --out-${NO_FUTURE}format colored-line-number --verbose --modules-download-mode=vendor] in [] ...
  pkg/metrics/bpf.go:10: File is not `goimports`-ed with -local github.com/cilium/cilium/ (goimports)
  	"github.com/prometheus/client_golang/prometheus"
  	"github.com/sirupsen/logrus"
  	"golang.org/x/sync/singleflight"
```

**Review by ti-mo — APPROVED** [review]
Clever, just keep in mind this will miss programs that don't refer to maps, if we ever add those. For now, this will do. :+1:

**Comment by mvisonneau** [comment]
thanks, I've just rebased, hope that all of tests will go through this time! 🤞

**Comment by joestringer** [comment]
A variety of tests have failed, I suggest triaging them to identify any potential links to the PR. You are of course welcome to rebase and retrigger CI again, but if there is an underlying issue in the PR itself then that course of action will have no impact.

Until the PR is passing CI, I'll mark this as draft. Feel free to reach out on Slack if you would like to discuss #testing issues or you need someone to retrigger the full CI. When CI is passing you can mark it "ready for review".

**Comment by mvisonneau** [comment]
Apologies for not having thoroughly followed-up on this! Seems like CI tests are passing now although we likely need a `/test` from a maintainer to run all of them I assume.

**Pull request merged**
